src/java.base/share/classes/java/lang/foreign/SegmentAllocator.java

```diff
@@ -257,89 +257,171 @@ default MemorySegment allocateFrom(AddressLayout layout, MemorySegment value) {
+     * @throws IndexOutOfBoundsException if {@code sourceOffset > source.byteSize() - (elementCount * sourceElementLayout.byteSize())}.
+     * @throws IndexOutOfBoundsException if either {@code sourceOffset} or {@code elementCount} are {@code < 0}.
+     */
+    @ForceInline
+    default MemorySegment allocateFrom(ValueLayout elementLayout, MemorySegment source,
+                                       ValueLayout sourceElementLayout, long sourceOffset, long elementCount) {
+        Objects.requireNonNull(source);
+        Objects.requireNonNull(sourceElementLayout);
+        Objects.requireNonNull(elementLayout);
+        MemorySegment dest = allocateNoInit(elementLayout, elementCount);
+        MemorySegment.copy(source, sourceElementLayout, sourceOffset, dest, elementLayout, 0, elementCount);
+        return dest;
+    }
+
     /**
      * Allocates a memory segment with the given layout and initializes it with the given byte elements.
-     * @implSpec the default implementation for this method calls {@code this.allocate(layout, array.length)}.
+     * @implSpec the default implementation for this method is equivalent to the following code:
+     * {@snippet lang = java:
+     * this.allocateFrom(layout, MemorySegment.ofArray(array),
+     *                   layout.withOrder(ByteOrder.nativeOrder()),
+     *                   0, array.length)
+     *}
      * @param elementLayout the element layout of the array to be allocated.
      * @param elements the byte elements to be copied to the newly allocated memory block.
      * @return a segment for the newly allocated memory block.
+     * @throws IllegalArgumentException if {@code elementLayout.byteAlignment() > elementLayout.byteSize()}.
      */
+    @ForceInline
     default MemorySegment allocateFrom(ValueLayout.OfByte elementLayout, byte... elements) {
-        return copyArrayWithSwapIfNeeded(elements, elementLayout);
+        return allocateFrom(elementLayout, MemorySegment.ofArray(elements),
+                elementLayout.withOrder(ByteOrder.nativeOrder()), 0, elements.length);
     }
 
     /**
      * Allocates a memory segment with the given layout and initializes it with the given short elements.
-     * @implSpec the default implementation for this method calls {@code this.allocate(layout, array.length)}.
+     * @implSpec the default implementation for this method is equivalent to the following code:
+     * {@snippet lang = java:
+     * this.allocateFrom(layout, MemorySegment.ofArray(array),
+     *                   layout.withOrder(ByteOrder.nativeOrder()),
+     *                   0, array.length)
+     *}
      * @param elementLayout the element layout of the array to be allocated.
      * @param elements the short elements to be copied to the newly allocated memory block.
      * @return a segment for the newly allocated memory block.
+     * @throws IllegalArgumentException if {@code elementLayout.byteAlignment() > elementLayout.byteSize()}.
```

Comment:
I've added this to all the array-accepting allocateFrom, as I think this will be thrown by the underlying copy op. There are a lot of other exceptions that memory copy could throw, in the following conditions:
1. layout size * element count is too big
2. offset/elementCount < 0
3. the interval (offset, size) falls outside the spatial bounds of the segment
4. access occurs from a thread other than the thread owning segment
5. access occurs when segment scope is no longer alive
6. start of segment is not aligned according to provided layout

Now, I think 1-5 are impossible because:
1. number of bytes is always derived from an array size, no overflow is possible
2. default impl guarantees positive lengths/offsets
3. access is provably in bounds (from 0 to array size)
4. array segment can be accessed from any thread
5. array segment is always alive

As for (6), I think right now it can happen, but I think we should probably rectify the implementation of the array-accepting `allocateFrom` to pass `JAVA_SHORT`, `JAVA_CHAR`, ... as the "source layout element" rather than just passing a modified version of the input destination layout. If we do that, then access would be aligned by construction.

Reply:
Okay, I agree with your last statement. We seem to be on the same page there.